zoo/pooltool/sum_to_three/config/sum_to_three_image_config.py

```diff
@@ -0,0 +1,138 @@
+K = 20  # num_of_sampled_actions
+num_simulations = 50
+update_per_collect = None
+batch_size = 256
+max_env_step = int(5e6)
+reanalyze_ratio = 0.0
+eval_freq = 1000
+# ==============================================================
+# end of the most frequently changed config specified by the user
+# ==============================================================
+
+render_config = RenderConfig.from_json(render_config_path)
+
+sumtothree_cont_sampled_efficientzero_config = dict(
+    exp_name=f"data_pooltool_ctree/sumtothree_image_sampled_efficientzero_k{K}_ns{num_simulations}_upc{update_per_collect}_rr{reanalyze_ratio}_seed0",
+    env=dict(
+        env_name="PoolTool-SumToThree",
+        env_type="not_board_games",
+        render_config_path=render_config_path,
+        observation_type="image",
+        collector_env_num=collector_env_num,
+        evaluator_env_num=evaluator_env_num,
+        n_evaluator_episode=evaluator_env_num,
+        manager=dict(
+            shared_memory=False,
+        ),
+    ),
+    policy=dict(
+        model=dict(
+            image_channel=render_config.channels,
+            observation_shape=render_config.observation_shape,
+            downsample=True,
+            action_space_size=2,
+            continuous_action_space=True,
+            categorical_distribution=False,
+            num_of_sampled_actions=K,
+            sigma_type="conditioned",
+            model_type="conv",
+            self_supervised_learning_loss=True,
+            res_connection_in_dynamics=True,
+            norm_type="BN",
+            # -------------------------
+            num_res_blocks=1,
+            num_channels=16,
+            fc_value_layers=[8],
+            fc_policy_layers=[8],
+        ),
+        cuda=True,
+        env_type="not_board_games",
+        game_segment_length=200,
+        update_per_collect=update_per_collect,
+        batch_size=batch_size,
+        # Optimizer
+        optim_type="Adam",
+        lr_piecewise_constant_decay=False,
+        learning_rate=0.003,
+        # NOTE: for continuous gaussian policy, we use the policy_entropy_loss as in the original Sampled MuZero paper.
+        policy_entropy_loss_weight=5e-3,
+        num_simulations=num_simulations,
+        reanalyze_ratio=reanalyze_ratio,
+        n_episode=n_episode,
+        eval_freq=eval_freq,
+        # the size/capacity of replay_buffer, in the terms of transitions.
+        replay_buffer_size=int(1e5),
+        collector_env_num=collector_env_num,
+        evaluator_env_num=evaluator_env_num,
+    ),
+)
+sumtothree_cont_sampled_efficientzero_config = EasyDict(
+    sumtothree_cont_sampled_efficientzero_config
+)
+main_config = sumtothree_cont_sampled_efficientzero_config
+sumtothree_cont_sampled_efficientzero_create_config = dict(
+    env=dict(
+        type="pooltool_sumtothree",
+        import_names=["zoo.pooltool.sum_to_three.envs.sum_to_three_env"],
+    ),
+    env_manager=dict(type="subprocess"),
```

Comment:
When running on my Linux server, no errors occurred. Could you please tell me what your system and Python versions are?

Reply:
I'm running on MacOS with the following system:

![image](https://github.com/opendilab/LightZero/assets/8688665/9cbd1859-a9f0-4f9b-ac89-d7607d43e82d)

Python version is:

```
▶▶ python
Python 3.10.14 (main, Mar 21 2024, 11:24:58) [Clang 14.0.6 ] on darwin
```

Pygame version is:

```
>>> import pygame
pygame 2.5.2 (SDL 2.28.3, Python 3.10.14)
Hello from the pygame community. https://www.pygame.org/contribute.html
```

Here is the full log:

[log.txt](https://github.com/user-attachments/files/15946325/log.txt)

Reply:
I just tried with a Python 3.8.19 environment and got the same error.

Reply:
Hello, I will be conducting tests on my local macOS. Thank you very much for providing the information.

Reply:
I noticed that if I take this `pygame.init()` call and move it outside of the class to the top of the module, the error is avoided.:

https://github.com/ekiefl/LightZero/blob/de799f3fc09e1a485215d3ec35416e297844b52e/zoo/pooltool/image_representation.py#L240

A side-effect is now the following warnings:

```
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
/Users/evan/anaconda3/envs/lightzero38/lib/python3.8/site-packages/ding/envs/env_manager/subprocess_env_manager.py:513: UserWarning: Cannot set number of intraop threads after parallel work has started or after set_num_threads call when using native parallel backend (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/ParallelNative.cpp:228.)
  torch.set_num_threads(1)
```

I'm not very confident with subprocess, so I'm not sure if we should be worried to see these warnings. But when I run `htop` I can see that the subprocesses are running.

One thing that worries me (I don't know if this is something to be worried about) is if `pygame.init()` is called outside the class, do all the subprocesses write to the same screen? If so, there will be race conditions for reading and writing to the pygame rendering.

If you aren't sure either, I can try and spend some time to make sure there are no race conditions occurring when `pygame.init()` is moved to the top of the module.

Reply:
OK my inexperience with pygame is showing 😄

https://www.pygame.org/docs/ref/pygame.html#pygame.init

I think there is no good reason to keep the pygame.init() call inside the class, since it just it just imports modules for pygame.

As for race conditions, I think the best way to prove there are no race conditions would be to show that the agents learn.

Reply:
I also conducted tests on a local macOS environment. By moving `pygame.init()` to the beginning of the class definition, the error we encountered previously can be avoided. Let's handle the issue using this method. Next, I will use the subprocess with the corresponding config to verify the performance. Thank you!